Modify rule S3229: Complete the description with bases; add compliant solution (#575)
SonarSource/rspec: rules/S3229/cfamily/rule.adoc

Class members are initialized in the order in which they are declared in the class,
not the order in which they appear in the class initializer list.
To avoid errors caused by order-dependent initialization,
the order of members in the initialization list should match the order in which members are declared in a class.

Base parts of the object are initialized before direct members, in the order they are listed in the inheritance list,

> and virtual bases are initialized first.

Reviewer: Given the next paragraph, I'm not sure this one brings any value. It could be removed I think.

Author: Indeed. Removed.

> The complete oprder, as described https://en.cppreference.com/w/cpp/language/constructor#Initialization_order[here]:

Reviewer: ```suggestion
The initialisation order, as described https://en.cppreference.com/w/cpp/language/constructor#Initialization_order[here], is:
```

Reviewer: or _initialization_ if American English is prefered, I don't know.

. If the constructor is for the most-derived class, virtual bases are initialized in the order in which they appear in depth-first left-to-right traversal of the base class declarations (left-to-right refers to the appearance in base-specifier lists)
. Then, direct bases are initialized in left-to-right order as they appear in this class's base-specifier list

> . Then, non-static data members are initialized in order of declaration in the class definition.

Reviewer: Github preview doesn't display this as a list. It may be better to use `*` to create a list here.

Author: That's because of the missing empty line before the list. I use `.` for a list to make it ordered (`*` creates an unordered list).

Reviewer: Ah, thanks for the explanation! :-)

== Noncompliant Code Example

----
#include <iostream>

struct A {
  A(int num) {

>     std::cout <<"A(num = " << num << <<")" <<std::endl;

Reviewer: ```suggestion
    std::cout << "A(num = " << num << ")" << std::endl;
```

  }
};

struct B {
  int b;
};

> class C :public A, B {

Reviewer: ```suggestion
class C : public A, B {
```

public:
  int x;
  int y;

  C(int i) : B{i}, A{b}, y(i), x(y + 1) { }  // Noncompliant
};

int main() {
  C c(1); // prints A(num = 0)

>   std::cout << c.x << " " << c.y << std::endl;  // prints 1 1

Reviewer: I actually wonder, is it well defined and always prints those values? I have the feeling that this (accessing members before they are initialised) is actually undefined behaviour... https://godbolt.org/z/41Gcxe7Ee seems to "confirm" this, although it would be more convincing to have the wording from the spec instead of an example but that's not the point of this rule.

It may be better to rephrase the comments to say it's UB and may print unexpected values instead of "1, 2, 1".

Author: Aggreed, rephrased.

}
----

== Compliant Solution

----
#include <iostream>

struct A {
  A(int num) {

>     std::cout <<"A(num = " << num << <<")" <<std::endl;

Reviewer: ```suggestion
    std::cout << "A(num = " << num << ")" << std::endl;
```

  }
};

struct B {
  int b;
};

> class C :public A, B {

Reviewer: ```suggestion
class C : public A, B {
```

public:
  int x;
  int y;

>   C(int i) : A{i}, B{i}, x(i + 1), y(i) { }  // Noncompliant

Reviewer: ```suggestion
  C(int i) : A{i}, B{i}, x(i + 1), y(i) { }
```
